interfaces/builtin/common_files.go

```diff
@@ -0,0 +1,162 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+type commonFilesInterface struct {
+	commonInterface
+
+	apparmorHeader    string
+	extraPathValidate func(string) error
+}
+
+// filesAAPerm can either be files{Read,Write} and converted to a string
+// expands into the right apparmor permissions for the files interface.
+type filesAAPerm int
+
+const (
+	filesRead filesAAPerm = iota
+	filesWrite
+)
+
+func (a filesAAPerm) String() string {
+	switch a {
+	case filesRead:
+		return "rk,"
+	case filesWrite:
+		return "rwkl,"
+	}
+	panic(fmt.Sprintf("invalid perm: %d", a))
+}
+
+func formatPath(ip interface{}) (string, error) {
+	p, ok := ip.(string)
+	if !ok {
+		return "", fmt.Errorf("%[1]v (%[1]T) is not a string", ip)
+	}
+	prefix := ""
+	if strings.HasPrefix(p, "$HOME") && strings.Count(p, "$HOME") == 1 {
+		p = strings.Replace(p, "$HOME", "@{HOME}", 1)
+		prefix = "owner "
+	}
+	p += "{,/,/**}"
+
+	return fmt.Sprintf("%s%q", prefix, filepath.Clean(p)), nil
+}
+
+func allowPathAccess(buf *bytes.Buffer, perm filesAAPerm, paths []interface{}) error {
+	for _, rawPath := range paths {
+		p, err := formatPath(rawPath)
+		if err != nil {
+			return err
+		}
+		fmt.Fprintf(buf, "%s %s\n", p, perm)
+	}
+	return nil
+}
+
+func (iface *commonFilesInterface) validatePaths(attrName string, paths []interface{}) error {
+	for _, npp := range paths {
+		np, ok := npp.(string)
+		if !ok {
+			return fmt.Errorf("%q must be a list of strings", attrName)
+		}
+		if err := iface.validateSinglePath(np); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (iface *commonFilesInterface) validateSinglePath(np string) error {
+	if strings.HasSuffix(np, "/") {
+		return fmt.Errorf(`%q cannot end with "/"`, np)
+	}
+	p := filepath.Clean(np)
+	if p != np {
+		return fmt.Errorf("%q must be clean", np)
+	}
+	if strings.Contains(p, "~") {
+		return fmt.Errorf(`%q contains invalid "~"`, p)
+	}
+	if err := apparmor.ValidateNoAppArmorRegexp(p); err != nil {
+		return err
+	}
+
+	// extraPathValidation must be implemented by the interface
+	// that build on top of the abstract commonFilesInterface
+	if iface.extraPathValidate == nil {
+		panic("extraPathValidate must be set when using the commonFilesInterface")
+	}
+	if err := iface.extraPathValidate(np); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (iface *commonFilesInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	hasValidAttr := false
+	for _, att := range []string{"read", "write"} {
+		if _, ok := plug.Attrs[att]; !ok {
+			continue
+		}
+		il, ok := plug.Attrs[att].([]interface{})
+		if !ok {
+			return fmt.Errorf("cannot add %s plug: %q must be a list of strings", iface.name, att)
```

Comment:
Should the string "cannot add foo plug:" be a part of an error message one level above this?